File: tp_algo_java/src/Compression.java
import java.util.ArrayList;
import java.util.HashMap;

public class Compression {
    private int m;
    private ArrayList<Byte> tab;
    private Binaire binaire;
    private ArrayList<HashMap<HeaderSequence,Integer>> memIteration;
    private ArrayList<Integer> memCout;
    private ArrayList<Integer> memSignificatif;

    int cpt = 0;

    public Compression(String fichier){
        this.binaire = new Binaire(fichier);

        this.m = binaire.getTaille();
        this.tab = binaire.getPixels();
        this.memIteration = new ArrayList<HashMap<HeaderSequence,Integer>>();
        this.memCout = new ArrayList<Integer>();
        this.memSignificatif = new ArrayList<Integer>();
        this.initMemorisation();
    }

    private void initMemorisation(){
    	memCout.clear();
    	memIteration.clear();
        memSignificatif.clear();

        memCout.add(Integer.MAX_VALUE);
        memIteration.add(new HashMap<HeaderSequence,Integer>());
        for(int i =0; i < m; i++){
            memCout.add(Integer.MAX_VALUE);
            memIteration.add(new HashMap<HeaderSequence,Integer>());
            memSignificatif.add(binaire.nbBits(tab.get(i)));
        }
    }

    public int c_recursif(){
    	this.initMemorisation();
        return cout_recursif(0,0,0);
    }

    private int cout_recursif(int i, int n, int b){
        HeaderSequence h = new HeaderSequence(n,b);

        //on a deja parcouru cet arbre
        if(memIteration.get(i).containsKey(h)){
            //System.out.println(memIteration.get(i));
            return memIteration.get(i).get(h);
        }

        //condition d'arret
        if(m <= i) {
            memIteration.get(i).put(h, 0);
            //System.out.println("fin " + b  + "   " + n);
        }
        else{
            int aj = memSignificatif.get(i);

            //si la séquence est remplie ou qu'il n'y en a pas :
            if(n == 0 || n >= 255){
                memIteration.get(i).put(h, new Integer(cout_recursif(i + 1, 1, aj) + 11 + aj));
            }
            //si le pixel est parfait pour la séquence :
            else if(b == aj){
                memIteration.get(i).put(h, new Integer(cout_recursif(i + 1, n + 1, aj) + aj));
            }
            //si le pixel est plus petit :
            else if(b > aj){
                int garde = cout_recursif(i+1, n+1, b) + b;
                int ferme = cout_recursif(i+1, 1, aj) + 11 + aj;
                memIteration.get(i).put(h, new Integer(Math.min(garde, ferme)));
            }
            //si le pixel est plus grand :
            else{
                int garde = cout_recursif(i+1, n+1, aj) + aj + (n-1)*(aj-b);
                int ferme = cout_recursif(i+1, 1, aj) + 11 + aj;
                memIteration.get(i).put(h, new Integer(Math.min(garde, ferme)));
            }
        }

        return memIteration.get(i).get(h);
    }
    
    public int c_iteratif(){
    	return cout_iteratif();
    }


    private int cout_iteratif() {
    	
    	this.memCout.set(this.m, 0);
    	this.memIteration.get(this.m).put(new HeaderSequence(0, 0), 0);
    	
    	for(int i=this.m-1; i>-1; i--){
    		HashMap<HeaderSequence,Integer> precedents = this.memIteration.get(i+1);
    		int a = memSignificatif.get(i);
    		
    		for (HeaderSequence k : precedents.keySet()){
    			int n = k.getN();
    			int b = k.getB();
    			int coutPrec = precedents.get(k);
    			
    			if ( n>=255 || n==0){
    				int newCout = coutPrec + 11 + a;
                    int newB = a;
                    int newN = 1;
                    
                    HeaderSequence h = new HeaderSequence(newN, newB);
                    if (this.memIteration.get(i).containsKey(h))
                        this.memIteration.get(i).put(h,Math.min(this.memIteration.get(i).get(h),newCout));
                    else
                        this.memIteration.get(i).put(h,newCout); 
    			}
    			else if(a == b){
    			    int newCout = coutPrec + a;
                    int newB = a;
                    int newN = n + 1;
                    
                    HeaderSequence h = new HeaderSequence(newN, newB);
                    if (this.memIteration.get(i).containsKey(h))
                        this.memIteration.get(i).put(h,Math.min(this.memIteration.get(i).get(h),newCout));
                    else
                        this.memIteration.get(i).put(h,newCout); 

    			}
    			else{
    				if (a<b){
    					int garde = coutPrec + b;
    					int ferme = coutPrec + a + 11;

    					int newCoutGarde = garde;
    					int newBGarde = b;
    					int newNGarde = n + 1;
    					int newCoutFerme = ferme;
    					int newBFerme = a;
    					int newNFerme = 1;
    					
    					HeaderSequence hGarde = new HeaderSequence(newNGarde, newBGarde);
    					HeaderSequence hFerme = new HeaderSequence(newNFerme, newBFerme);
    					
    					if (this.memIteration.get(i).containsKey(hGarde))
                            this.memIteration.get(i).put(hGarde,Math.min(this.memIteration.get(i).get(hGarde), newCoutGarde));
                        else
                            this.memIteration.get(i).put(hGarde, newCoutGarde);

                        if (this.memIteration.get(i).containsKey(hFerme))
                        	this.memIteration.get(i).put(hFerme,Math.min(this.memIteration.get(i).get(hFerme), newCoutFerme));
                        else
                        	this.memIteration.get(i).put(hFerme, newCoutFerme);
    				}
    				else if (a>b){
    					int garde = coutPrec + a + n*(a-b);
						int ferme = coutPrec + a + 11;

						int newCoutGarde = garde; 
						int newBGarde = a;
						int newNGarde = n + 1;
						int newCoutFerme = ferme;
						int newBFerme = a;
						int newNFerme = 1;
                        
                        HeaderSequence hGarde = new HeaderSequence(newNGarde, newBGarde);
    					HeaderSequence hFerme = new HeaderSequence(newNFerme, newBFerme);
    					
    					if (this.memIteration.get(i).containsKey(hGarde))
                            this.memIteration.get(i).put(hGarde,Math.min(this.memIteration.get(i).get(hGarde), newCoutGarde));
                        else
                            this.memIteration.get(i).put(hGarde, newCoutGarde);

                        if (this.memIteration.get(i).containsKey(hFerme))
                        	this.memIteration.get(i).put(hFerme,Math.min(this.memIteration.get(i).get(hFerme), newCoutFerme));
                        else
                        	this.memIteration.get(i).put(hFerme, newCoutFerme);
    					
    				}
    			}
                         
    		}
    		this.setMinimumPixel(i);
    	}
    	
		return this.memCout.get(0);
	}

	private void setMinimumPixel(int i) {
		Integer minvalue = Integer.MAX_VALUE;
		
		for (Integer value : this.memIteration.get(i).values()){
			if (minvalue > value){
				minvalue = value;
			}
		}
		
		this.memCout.set(i, minvalue.intValue());
	}

	public int getM(){
        return m;
    }

        /*
        public int cout(int i, int n, int b){
        HeaderSequence h = new HeaderSequence(n,b);

        //on a deja parcouru cet arbre
        if(memIteration.get(i) != null && memIteration.get(i).containsKey(h)){
            return memIteration.get(i).get(h);
        }

        HashMap<HeaderSequence,Integer> value;
        if(memIteration.get(i) != null)
            value = memIteration.get(i);
        else{
            value = new HashMap<HeaderSequence,Integer>();
            memIteration.set(i, value);
        }

        //condition d'arret
        if(m <= i) {
            value.put(h, new Integer(0));
        }
        else{
            int aj = binaire.nbBits(tab.get(i));

            //si la séquence est remplie ou qu'il n'y en a pas :
            if(n == 0 || n >= 255){
                value.put(h, new Integer(cout(i+1, 1, aj) + 11 + aj));
            }
            //si le pixel est parfait pour la séquence :
            else if(b == aj){
                    value.put(h, new Integer(cout(i+1, n+1, aj) + aj));
            }
            //si le pixel est plus petit :
            else if(b > aj){
                int garde = cout(i+1, n+1, b) + b;
                int ferme = cout(i+1, 1, aj) + 11 + aj;
                value.put(h, new Integer(Math.min(garde,ferme)));
            }
            //si le pixel est plsu grand :
            else{
                int garde = cout(i+1, n+1, aj) + aj + (n-1)*(aj-b);
                int ferme = cout(i+1, 1, aj) + 11 + aj;
                value.put(h, new Integer(Math.min(garde,ferme)));
            }
        }
         */
}

File: Python/compression.py
import struct
import sys
from binaire import GestionBinaire

class Compression:
    
    def __init__(self, fichier):        
        self.filePath = fichier        
        
        # helper binaire
        self.binaire = GestionBinaire()
        self.binaire.readFile(fichier)
        
        # Taille de l'image carre en largeur/longueur
        self.m = self.binaire.getTaille()
        
        # Tableau linearise
        self.tab = self.binaire.getPixels()
        
        # Tableau memoisation
        self.memCout = []
        self.memIteration = []
        
        #Initialisation des tableaux de memoisation
        self.initMem()
        
    def initMem(self):
        for i in range(0,(self.m**2)+1):
            self.memCout.append(-1)
            self.memIteration.append({})
    
    def c(self):
        print "Algorithme iteratif :"
        print "fichier de depart : ",(self.m**2), " octets soit ", (self.m**2)*8, " bits"
        cout = self.__cout_iteratif()
        print "fichier apres compression : ",cout/8, " octets soit ", cout, " bits"
        self.solution()
    
    def __cout_iteratif(self):
        
        #Initialisation de la derniere case du tableau (juste apres le dernier pixel)
        self.memCout[self.m**2] = 0
        self.memIteration[self.m**2][0,0] = 0

        # Iteration sur les pixels
        for i in range((self.m**2)-1,-1, -1):
            
            precedents = self.memIteration[i+1]
            a = self.binaire.nbBits(self.tab[i])
            
            # Iteration sur les combinaisons (i,n,b) de la case precedemment calculee
            for keys, coutPrec in precedents.items() :
                n = keys[0]
                b = keys[1]
                
                # sequence rempli ou aucune sequence
                if n >= 255 or n == 0 :                
                    
                    if (1,a) in self.memIteration[i] :
                        self.memIteration[i][1,a] = min(self.memIteration[i][1,a],coutPrec + 11 + a)
                    else :
                        self.memIteration[i][1,a] = coutPrec + 11 + a 

                    
                # Si le pixel est parfait pour la sequence :
                elif a == b :
                    
                    if (n+1,a) in self.memIteration[i] :
                        self.memIteration[i][n+1,a] = min(self.memIteration[i][n+1,a],coutPrec + a)
                    else :
                        self.memIteration[i][n+1,a] = coutPrec + a
 
                else:
                        
                    # Si le pixel est plus petit :
                    if a < b :                
                        garde = coutPrec + b
                        ferme = coutPrec + a + 11

                        if (n+1,b) in self.memIteration[i] :
                            self.memIteration[i][n+1,b] = min(self.memIteration[i][n+1,b],garde)
                        else :
                            self.memIteration[i][n+1,b] = garde 

                        if (1,a) in self.memIteration[i] :
                            self.memIteration[i][1,a] = min(self.memIteration[i][1,a],ferme)
                        else :
                            self.memIteration[i][1,a] = ferme 
                    
                    # Si le pixel est plus grand :
                    elif a > b :
                        garde = coutPrec + a + n*(a-b)
                        ferme = coutPrec + a + 11

                        if (n+1,a) in self.memIteration[i] :
                            self.memIteration[i][n+1,a] = min(self.memIteration[i][n+1,a],garde)
                        else :
                            self.memIteration[i][n+1,a] = garde 

                        if (1,a) in self.memIteration[i] :
                            self.memIteration[i][1,a] = min(self.memIteration[i][1,a],ferme)
                        else :
                            self.memIteration[i][1,a] = ferme 

            self.memCout[i] = min(value for value in self.memIteration[i].values())

        return self.memCout[0]
    
    def solution(self):
        nPrec = 0
        bPrec = 0
        # Initialise n et b de la premiere case
        for keys, coutPrec in self.memIteration[0].items():
            if self.memCout[0] == coutPrec :
                nPrec = keys[0]
                bPrec = keys[1]
        
        # Iteration sur le tableau des combinaisons
        for i in range(1,(self.m**2)+1):
            #Iteration sur les combinaisons
            for keys, coutPrec in self.memIteration[i].items():
                n = keys[0]
                b = keys[1]
                
                if i == 5619 and n == 5 and b==8 :
                    print self.memIteration[i]
                # si on doit ouvrir une nouvelle sequence
                if nPrec == 1:
                    if (self.memCout[i-1] != coutPrec + 11 + bPrec):
                        del self.memIteration[i][keys]
                # si le n ne correspond pas, on supprime les mauvaises valeurs
                elif n != nPrec-1 or b > bPrec :
                    del self.memIteration[i][keys]
            
            #Iteration sur les combinaisons qui correspondent au n de la meilleur solution
            self.memCout[i] = min(value for value in self.memIteration[i].values())
            for keys, coutPrec in self.memIteration[i].items():
                if self.memCout[i] != coutPrec:
                    del self.memIteration[i][keys]
                else:
                    nPrec = keys[0]
                    bPrec = keys[1]
                    
    def __cout_recursif(self, i, n, b):

        # On a deja parcouru cette arbre
        if (n,b) in self.memIteration[i] :
            return self.memIteration[i][n,b]

        # Condition d'arret
        if (self.m**2) == i :
            self.memIteration[i][n,b] = 0
        else :
            aj = self.binaire.nbBits(self.tab[i])

            if n == 0 or n >= 255 :
                self.memIteration[i][n,b] = self.__cout(i+1, 1, aj) + 11 + aj

            # Si le pixel est parfait pour la sequence :
            elif b == aj:
                self.memIteration[i][n,b] = aj + self.__cout(i+1, n+1, aj)

            # Si le pixel est plus petit : 
            elif b > aj:

                garde = self.__cout(i+1, n+1, b) + b
                ferme = self.__cout(i+1, 1, aj) + aj + 11
                self.memIteration[i][n,b] = min(garde, ferme)

            # Si le pixel est plus grand :
            else :

                garde = self.__cout(i+1, n+1, aj) + aj + (n-1)*(aj-b)
                ferme = self.__cout(i+1, 1, aj) + aj + 11
                self.memIteration[i][n,b] = min(garde, ferme)

        if self.memCout[i] == -1 or self.memIteration[i][n,b] < self.memCout[i] :
            self.memCout[i] = self.memIteration[i][n,b]

        return self.memIteration[i][n,b]
            
            
    def uncompresse(self):
		print "decompression initialisee"
		self.tab = self.binaire.readFileToUncompress(self.filePath)
		tab = self.tab
		result = []
		
		i=0
		#parcours de tous les bytes du fichier a decompresser

		#recuperation des variables dans l'entete du segment
		#n = struct.unpack('<B',octet)
		n = tab[i]
		i+=1
		b = tab[i] >> 5
		nbBitRestant = 5
		
		while i < len(tab) and n > 0:
			if n == 0:
				#recuperation des variables dans l'entete du segment
				n = tab[i]
				i+=1
				b = tab[i] >> 5
				nbBitRestant = 5
				
				if nbBitRestant == 8:
					n = tab[i]
					nbBitRestant = 8
				else:
					octet = tab[i]
					i+=1
					if(i >= len(tab)):
						break
					nbBitEnPlus = 8 - nbBitRestant
					
					octet1 = octet << nbBitEnPlus
					octet2 = tab[i] >> (8-nbBitEnPlus)
					n = octet1 or octet2
					nbBitRestant = 8-nbBitEnPlus

				#Recupere le b
				if(i >= len(tab)):
					break
				if nbBitRestant == 3:
					b = tab[i]
					nbBitRestant = 8
				elif nbBitRestant > 3:
					octetUnCompress = tab[i] >> nbBitRestant-3
					nbBitRestant = nbBitRestant-3
				else:
					octet = tab[i]
					i+=1
					if(i >= len(tab)):
						break
					nbBitEnPlus = 3 - nbBitRestant
					
					octet1 = octet << nbBitEnPlus
					octet2 = tab[i] >> (8-nbBitEnPlus)
					b = octet1 or octet2
					nbBitRestant = 8-nbBitEnPlus

			
			else:
				#traitement des octets de la sequence
			
				#garde que les bits non traites de l'octet
				octet = tab[i] and ((2^nbBitRestant)-1)
				if nbBitRestant == b:
					octetUnCompress = octet
					nbBitRestant = 8
				elif nbBitRestant > b:
					octetUnCompress = octet >> nbBitRestant-b
					nbBitRestant = nbBitRestant-b
				else:
					i+=1
					if(i >= len(tab)):
						break
					
					nbBitEnPlus = b - nbBitRestant
					
					octet1 = octet << nbBitEnPlus
					octet2 = tab[i] >> (8-nbBitEnPlus)
					octetUnCompress = octet1 or octet2
					nbBitRestant = 8-nbBitEnPlus
			
				result.append(octetUnCompress)
				n -=1
			i +=1
			
		print "decompression terminee"
		print "creation du fichier decompresse"
		self.binaire.createUnCompressFile(self.filePath ,result)
		print "votre fichier est disponible"
        
    def shouldWeCompressTheFile(self):
        listCutFile = self.filePath.split('.')
        length = len(listCutFile)
        if listCutFile[length-1] == 'seg':
            return False
        else:
            return True
        
filePath = sys.argv[1]
print filePath
compression = Compression(filePath)
if(compression.shouldWeCompressTheFile()):
    compression.c()
else:
    compression.uncompresse()    
    
    
    
    
